### ISSUE TITLE
Support RPM install better with custom --prefix option.

### ISSUE THREAD

**Issue by ajoliveira** [elastic/elastic-agent#141]
**Describe the enhancement:**
For comparison, the Elasticsearch RPM file allows relocation of the `/usr` directory ONLY.

RPM supports the use of a '--prefix' flag (ex: `--prefix=/opt`) to change the installation path of files listed in the matching directory (if allowed).  For example, the following rpm command output - ex: Relocations:
```
[elastic@localhost ~]$ rpm -qip metricbeat-7.15.1-x86_64.rpm
Name        : metricbeat
Version     : 7.15.1
Release     : 1
Architecture: x86_64
Install Date: (not installed)
Group       : default
Size        : 175184387
License     : Elastic-License
Signature   : RSA/SHA512, Thu 07 Oct 2021 08:35:04 PM EDT, Key ID d27d666cd88e42b4
Source RPM  : metricbeat-7.15.1-1.src.rpm
Build Date  : Thu 07 Oct 2021 06:26:04 PM EDT
Build Host  : 030a0d86dfd1
Relocations : /
Packager    : <@030a0d86dfd1>
Vendor      : Elastic
URL         : https://www.elastic.co/beats/metricbeat
Summary     : Metricbeat is a lightweight shipper for metrics.
Description :
Metricbeat is a lightweight shipper for metrics.
```

In the case of metricbeat, the 3 top level directories that are used for install are `/etc`, `/lib`, and `/usr`.  Since the allowed relocation is the top-level root directory `/`, all 3 directories will be moved since they are sub-directories of this directory.  This has an unfortunate side effect that if you are not able to install into `/usr` for example, you cannot change ONLY that directory.  Installing with the `--prefix=/opt/metricbeat` option:
`sudo rpm -ivh --prefix=/opt/metricbeat metricbeat-7.15.1-x86_64.rpm`

everything ends up in `/opt/metricbeat` with no files/directories in `/etc/` or `/lib` and no service is installed.  The workaround can be used to add the service and update configuration files, ex:
1) move files from `/opt/metricbeat/etc/*` to `/etc` and `/opt/metricbeat/lib/*` to `/lib`
     - or maybe a better option is to link them instead for a cleaner uninstall?
2) edit resulting `/lib/systemd/system/metricbeat.service`, prepending any references to `/usr/share` with `/opt/metricbeat` (ex: `/opt/metricbeat/usr/share/metricbeat`)
3) do same for resulting `/etc/init.d/metricbeat`
4) enable the metricbeat service for startup (ex: `sudo systemctl enable metricbeat`)
5) start/stop service as required

**Describe a specific use case for the enhancement or feature:**
In some cases, users are not allowed to (or it is recommended that they not) install any files to the `/usr` directory and would prefer to install to the `/opt` directory as an alternative - WHILE still allowing for proper installation relative to configuration files and startup scripts/services and not require manual steps to get working properly.

**Comment by nimarezainia** [comment]
@ajoliveira sorry for the delay. Was there a specific customer request for this? I understand the use case and why it could be important.

I would prefer to make this enhancement to elastic-agent. your thoughts?

**Comment by nimarezainia** [comment]
@ajoliveira pls let me know. At the moment ideally I'd like to concentrate on customer found enhancement requests

**Comment by ajoliveira** [comment]
@nimarezainia Yes, this was found as an issue at a customer installation.  Details were provided in direct a e-mail and while it was originally specific to metricbeat, I'll defer to you as to when and how we can implement this.  While not optimal, there is a workaround if needed in the meantime.

**Comment by pierrehilbert** [comment]
@nimarezainia should we spend some time on this one or is it something we want to only allow when installed from tar.gz?

**Comment by nimarezainia** [comment]
@pierrehilbert can the rpm/deb at all use a different install location?

**Comment by pierrehilbert** [comment]
Not currently, only possible when installing through CLI.

**Comment by nimarezainia** [comment]
@cmacknz what's the complexity in getting this done?

**Comment by ebeahan** [comment]
Changed the status from `Needs Tech Definition` to `Ready`. The RPM package format already supports `--prefix` to define a custom installation location. Whoever takes this on can either capture how it will work as a comment here or, if it's simpler, go straight to opening a PR with an implementation.

**Comment by michel-laterman** [comment]
I don't think that [fpm](https://github.com/jordansissel/fpm) (the tool we use to generate the RPM and DEB packages) has an easy way for us to define how the  (rpm) `--prefix` option is handled - especially as we use [quite a few scripts](https://github.com/elastic/elastic-agent/tree/main/dev-tools/packaging/templates/linux) to manage package installation.

However the `--prefix` path is accessible via the `$RPM_INSTALL_PREFIX` var in each of the scripts.
If no prefix is passed then `RPM_INSTALL_PREFIX=/`.

Currently after running a package installation the agent has files in:
```
/var/lib/elastic-agent    # majority of installation; binaries, components, etc. agent uses this as it's home dir for local storage
/etc/elastic-agent/       # configs - used by standalone agent
/usr/share/elastic-agent/ # NOTICE, README, LICENSE and a symlink in the bin dir
```

And creates:
`/usr/bin/elastic-agent` a script that runs `/usr/share/elastic-agent/bin/elastic-agent` (a symlink to `/var/lib/elastic-agent/...`), and has hardcoded references to other directories (`paths.home=/var/lib/elastic-agent`, `paths.config=/etc/elastic-agent`, `paths.logs=/var/log/elastic-agent`)

And `/lib/systemd/system/elastic-agent.service` that points to `/usr/bin/elastic-agent` and references `/etc/elastic-agent`.

If we were to use a prefix with our current layout, all files would be under the prefix. i.e., `--prefix=/opt/elastic-agent` would result in `/opt/elastic-agent/var/lib/elastic-agent`, `/opt/elastic-agent/etc/elastic-agent`, etc.

I think that as far as scope goes, we would need to add prefix handing for our install scripts (they all use hard-coded paths at the moment), and symlinks if a custom prefix is used, i.e.,
`/lib/systemd/system/elastic-agent.service -> /opt/elastic-agent/lib/systemd/system/elastic-agent.service` and `/usr/bin/elastic-agent -> /opt/elastic-agent/usr/bin/elastic-agent`
I can move forward with trying to make these adjustments function for RPM systems; we can use `sed` to edit the service file and entrypoint to adjust the hardcoded locations.
@nimarezainia, am I understanding the scope here correctly, or have I missed something? 

As a side note deb has the `--instdir` or `--root` args that seem to do something similar; but I have not done any investigation on what they appear as in the scripts.

**Comment by nimarezainia** [comment]
> [@nimarezainia](https://github.com/nimarezainia), am I understanding the scope here correctly, or have I missed something?

@michel-laterman no you got it. We want the same behaviour when --prefix is used on a tarball install. The user doesn't have access/permission to the default paths we use and should define an alternate path for the installation.

**Comment by jlind23** [comment]
cc @lucabelluccini , this probably deserves an entry in the next release summary.

**Comment by lucabelluccini** [comment]
Thanks @jlind23  for the heads up.
I'll make sure this will be in the awareness KB
